### PR TITLE
feat: add tson

### PR DESCRIPTION
**Pull request by skarab42**
Hello, I would like to propose my validation library :)
https://github.com/skarab42/tson

### CODE REVIEW

**Comment by moltar** [comment]
Getting:

>     Cannot find module '@skarab/tson' from 'cases/tson.ts'

Any ideas?

**Comment by skarab42** [comment]
Do you see it in the node_modules directory?

**Comment by moltar** [comment]
> Do you see it in the node_modules directory?

Not sure, this came back from GitHub Actions CI run :)

**Comment by skarab42** [comment]
Can you try locally?

**Comment by skarab42** [comment]
@moltar I'm not sure, but I don't see the install in the logs, I added the command to the workflow.

File: .github/workflows/pr.yml [hunk 1]
```diff
@@ -42,6 +42,8 @@ jobs:
       - name: install
         run: npm ci
 
+      - run: npm install
```
Comment (moltar) [inline .github/workflows/pr.yml:45]: But `npm ci` does install too?

File: package-lock.json [hunk 2]
```diff
@@ -1,9266 +1,8 @@
 {
   "name": "typescript-runtime-type-benchmarks",
   "version": "1.0.0",
-  "lockfileVersion": 2,
```
Comment (moltar) [inline package-lock.json:4]: Oh, lockfile has changed. I think older Node & NPM does not support v2?

**Comment by moltar** [comment]
Try rebuilding the lockfile with an older NPM version to go back to v1.

File: package-lock.json (hunk 2 above)
Comment (skarab42) [reply package-lock.json]: sorry i haven't used npm for a long time, i switched to yarn v1 very quickly and now i mainly use pnpm.

**Comment by skarab42** [comment]
@moltar Is there anything else I can do?

**Comment by skarab42** [comment]
what is strange is that there is no error at the installation

**Comment by moltar** [comment]
Sorry not ignoring you just a bit busy with a project delivery.

**Comment by moltar** [comment]
Maybe @hoeck has some ideas?

**Comment by hoeck** [comment]
Hi there,

looks like tson targets only newer node versions by explicitly not compiling down features like the `??`-operator. This is what I get when trying to import tson using node 12:

```
> process.version
'v12.16.1'
> require('@skarab/tson')
/home/erik/tmp/testtsoninstall/node_modules/@skarab/tson/build/cjs/type/function.js:7
    const argTypes = input[0] ?? [];
                               ^

Uncaught SyntaxError: Unexpected token '?'
```

Your [tsconfig](https://github.com/skarab42/tson/blob/620ead4c434d36a510aec74aef64cce90dd32e80/tsconfig.json#L4) seems to confirm that, [see also typescript docs about the `target` option](https://www.typescriptlang.org/tsconfig#target).

**Comment by skarab42** [comment]
Thanks for the investigation, I'll change that.

**Comment by skarab42** [comment]
I have made the change, I hope it will solve the issue.

**Comment by hoeck** [comment]
Great, lets see what the workflows say. Locally requiring @skarab/tson  1.4.3 on node 12 worked for me.

**Comment by skarab42** [comment]
It looks like the CI doesn't install the dependencies.

**Comment by hoeck** [comment]
I think it does install them. Running `npm test` locally I get the same error, regardless of the node version. Importing tson using ts-node works without problems now.

So it looks like an issue with ... Jest? I'm out of ideas but still trying to debug it now :sweat_smile:

**Comment by moltar** [comment]
I'm stumped as well. Just spent half an hour trying to figure out jest failure, and can't see why it would be failing.

Does it work in isolation in any other consuming project?

E.g. if starting a brand new NPM project, add tson and jest. Does it work?

**Comment by skarab42** [comment]
Don't waste any more time, the problem comes from tson. My build system is not great. I'm doing a new setup with tsup.

**Comment by hoeck** [comment]
> Don't waste any more time

LOL too late, got stuck in the tarpit already :laughing: 

Its a combination of Jest using its own code to resolve imported modules (jest-resolve) and your module having a package.json in your dist files at `node_modules/@skarab/tson/build/cjs/package.json`. When removing that file, the tests will run fine.

What is the purpose of that package.json? If this is some standard node module stuff then maybe we need a different jest-configuration or maybe its a jest bug?

**Comment by skarab42** [comment]
> What is the purpose of that package.json? If this is some standard node module stuff then maybe we need a different jest-configuration or maybe its a jest bug?

I took this (stupidly) from an old project.

The new realese uses tsup which produces a clean esm/cjs build (I hope). I tested on a clean install with jest and node 12 and it seems that v1.5.0 works.

**Comment by skarab42** [comment]
Ok, one more step, I see where the problem is, I'll fix it. Thanks for your patience.

**Comment by moltar** [comment]
> Its a combination of Jest using its own code to resolve imported modules (jest-resolve) and your module having a package.json in your dist files at `node_modules/@skarab/tson/build/cjs/package.json`. When removing that file, the tests will run fine.

Ah nice find!!

**Comment by skarab42** [comment]
Ok this time I hope it's the right one ;) 
I also commented on the import of `rulr` which breaks the test.

**Comment by skarab42** [comment]
Yhea! thanks a lot for your help, I really appreciate it :)

**Pull request merged**